### PR TITLE
RM-6942: DiagnosticMetadataAttribute BocListCellContents is not rendered when value is null

### PR DESCRIPTION
**Pull request by iam-mholle**
https://re-motion.atlassian.net/browse/RM-6942

### CODE REVIEW

**Comment by iam-mholle** [comment]
The issue is not yet resolved.

**Comment by iam-mholle** [comment]
2 remaining things to clarify:

> Also, rename BocValueColumnRendererBase Property "HasDiagnosticMetadata" to "HasContentAttribute" to better reflect the functionality of the property.

```csharp
// DiagnosticMetadataAttributesForObjectBinding.BocListColumnHasDiagnosticMetadata = "data-boclist-column-has-dma";
renderingContext.Writer.AddAttribute (
    DiagnosticMetadataAttributesForObjectBinding.BocListColumnHasDiagnosticMetadata,
    HasDiagnoticMetadata.ToString().ToLower());
```
I do think that the name is fine as-is.

> Also, add a check to the rendering of the DiagnosticMetadata BocListCellContents, which checks if the "HasContentAttribute" is set to true. TBD if the check should be an 'Assert' or just another 'if' condition

Is this needed? `BocCommandEnabledColumnRendererBase.RenderValueColumnCellText` does check if `RenderingFeatures.EnableDiagnosticMetadata` is true.

File: Remotion/ObjectBinding/Web.Development.WebTesting/ControlObjects/BocListControlObjectBase.cs [hunk 1]
```diff
@@ -456,10 +456,10 @@ protected override ElementScope GetLabeledElementScope ()
 
     private bool ColumnHasDiagnosticMetadata (ElementScope scope)
```
Comment (MichaelKetting) [inline Remotion/ObjectBinding/Web.Development.WebTesting/ControlObjects/BocListControlObjectBase.cs]: fix name

File: Remotion/ObjectBinding/Web/UI/Controls/BocListImplementation/Rendering/BocColumnRendererBase.cs [hunk 2]
```diff
@@ -126,15 +126,14 @@ void IBocColumnRenderer.RenderTitleCell (BocColumnRenderingContext renderingCont
...
+        var columnTitle = StringUtility.NullToEmpty (renderingContext.ColumnDefinition.ColumnTitleDisplayValue);
+        renderingContext.Writer.AddAttribute (DiagnosticMetadataAttributes.Content, HtmlUtility.StripHtmlTags (columnTitle));
 
         var oneBasedCellIndex = renderingContext.VisibleColumnIndex + 1;
         renderingContext.Writer.AddAttribute (DiagnosticMetadataAttributesForObjectBinding.BocListCellIndex, oneBasedCellIndex.ToString());
 
         renderingContext.Writer.AddAttribute (
-            DiagnosticMetadataAttributesForObjectBinding.BocListColumnHasDiagnosticMetadata,
+            DiagnosticMetadataAttributesForObjectBinding.BocListColumnHasContentAttribute,
             HasDiagnoticMetadata.ToString().ToLower());
```
Comment (MichaelKetting) [inline Remotion/ObjectBinding/Web/UI/Controls/BocListImplementation/Rendering/BocColumnRendererBase.cs]: Rename.

Comment (MichaelKetting) [reply Remotion/ObjectBinding/Web/UI/Controls/BocListImplementation/Rendering/BocColumnRendererBase.cs]: We have one more problem: in SimpleValueColumnRendererer, the content attribute is not rendered if the column is in edit mode. Am not sure how to best solve this. Maybe we just need to render the current value as attribute in SimpleValueColumnRenderer.RenderEditModeControl().

Comment (iam-mholle) [reply Remotion/ObjectBinding/Web/UI/Controls/BocListImplementation/Rendering/BocColumnRendererBase.cs]: @MichaelKetting  Would this be the desired outcome?

- Edit mode:
image

- Normal mode for reference:
image

Comment (iam-mholle) [reply Remotion/ObjectBinding/Web/UI/Controls/BocListImplementation/Rendering/BocColumnRendererBase.cs]: For the sake of completeness, that's what the nesting looks like with commands.

- Display name with command:
image

- Command (no `data-boclist-cell-contents` - makes sense):
image

File: Remotion/ObjectBinding/Web.UnitTests/UI/Controls/BocListImplementation/Rendering/BocSimpleColumnRendererTest.cs [hunk 3]
```diff
@@ -179,5 +212,37 @@ public void RenderEditModeControl ()
               0,
               List.ClientID + "_0_Title"));
     }
+
+    [Test]
+    public void RenderEditModeControl_HasDiagnosticMetadata ()
```
Comment (MichaelKetting) [inline Remotion/ObjectBinding/Web.UnitTests/UI/Controls/BocListImplementation/Rendering/BocSimpleColumnRendererTest.cs]: TestDiagnosticMetadataRenderingWithEditModeControl

File: Remotion/ObjectBinding/Web.UnitTests/UI/Controls/BocListImplementation/Rendering/BocSimpleColumnRendererTest.cs [hunk 4]
```diff
@@ -77,6 +76,40 @@ public void RenderBasicCell ()
       Html.AssertTextNode (textWrapper, "referencedObject1", 0);
     }
 
+    [Test]
+    public void RenderCell_WithEmptyDisplayName ()
```
Comment (MichaelKetting) [inline Remotion/ObjectBinding/Web.UnitTests/UI/Controls/BocListImplementation/Rendering/BocSimpleColumnRendererTest.cs]: TestDiagnosticMetadataRenderingWithEmptyDisplayName

File: Remotion/ObjectBinding/Web.UnitTests/UI/Controls/BocListImplementation/Rendering/BocSimpleColumnRendererTest.cs [hunk 5]
```diff
@@ -77,6 +76,40 @@ public void RenderBasicCell ()
...
+
+      var document = Html.GetResultDocument();
+      var td = Html.GetAssertedChildElement (document, "td", 0);
+      var span = Html.GetAssertedChildElement (td, "span", 0);
+      var textWrapper = Html.GetAssertedChildElement (span, "span", 0);
+      Html.AssertAttribute (textWrapper, DiagnosticMetadataAttributesForObjectBinding.BocListCellContents, string.Empty);
+    }
+
+    [Test]
+    public void RenderCell_WithNullAsContent ()
```
Comment (MichaelKetting) [inline Remotion/ObjectBinding/Web.UnitTests/UI/Controls/BocListImplementation/Rendering/BocSimpleColumnRendererTest.cs]: TestDiagnosticMetadataRenderingWithNullValue

**Comment by iam-mholle** [comment]
Added to master with 90fba06d10464a0da5865ef5d80e7c1b57df0790.

**Pull request closed without merging**